File: apps/sandbox-runtime/internal/tunnel/pty_channel.go
package tunnel

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"syscall"

	"github.com/coder/websocket"
	"github.com/mistlehq/mistle/apps/sandbox-runtime/internal/sessionprotocol"
)

const ptyOutputReadBufferBytes = 4096

type ptyControlAction string

const (
	ptyControlActionContinue     ptyControlAction = "continue"
	ptyControlActionDetachStream ptyControlAction = "detach-stream"
	ptyControlActionCloseSession ptyControlAction = "close-session"
)

func handlePTYConnectRequest(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	connectRequest connectRequest,
	activePTYSession *ptySession,
	relayResultCh chan<- activeTunnelStreamRelayResult,
) (*ptySession, *activeTunnelStreamRelay, error) {
	ptyConnectRequest, err := parsePTYConnectRequest(connectRequest.RawPayload)
	if err != nil {
		if writeErr := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
			Type:     sessionprotocol.MessageTypeStreamOpenError,
			StreamID: connectRequest.StreamID,
			Code:     connectErrorCodeInvalidConnectRequest,
			Message:  err.Error(),
		}); writeErr != nil {
			return activePTYSession, nil, fmt.Errorf("failed to write sandbox tunnel stream.open error: %w", writeErr)
		}
		return activePTYSession, nil, nil
	}

	if ptyConnectRequest.Channel.Session == sessionprotocol.PTYSessionModeCreate {
		if activePTYSession != nil && !activePTYSession.IsExited() {
			if writeErr := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
				Type:     sessionprotocol.MessageTypeStreamOpenError,
				StreamID: connectRequest.StreamID,
				Code:     connectErrorCodePTYSessionExists,
				Message:  "pty session already exists",
			}); writeErr != nil {
				return activePTYSession, nil, fmt.Errorf("failed to write sandbox tunnel stream.open error: %w", writeErr)
			}
			return activePTYSession, nil, nil
		}

		activePTYSession, err = startPTYSession(ptyConnectRequest)
		if err != nil {
			if writeErr := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
				Type:     sessionprotocol.MessageTypeStreamOpenError,
				StreamID: connectRequest.StreamID,
				Code:     connectErrorCodePTYSessionCreateFailed,
				Message:  err.Error(),
			}); writeErr != nil {
				return activePTYSession, nil, fmt.Errorf("failed to write sandbox tunnel stream.open error: %w", writeErr)
			}
			return activePTYSession, nil, nil
		}
	}

	if ptyConnectRequest.Channel.Session == sessionprotocol.PTYSessionModeAttach {
		if activePTYSession == nil || activePTYSession.IsExited() {
			if writeErr := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
				Type:     sessionprotocol.MessageTypeStreamOpenError,
				StreamID: connectRequest.StreamID,
				Code:     connectErrorCodePTYSessionUnavailable,
				Message:  "pty session is not available",
			}); writeErr != nil {
				return activePTYSession, nil, fmt.Errorf("failed to write sandbox tunnel stream.open error: %w", writeErr)
			}
			return activePTYSession, nil, nil
		}
	}

	if err := writeStreamOpenOK(ctx, tunnelConn, sessionprotocol.StreamOpenOK{
		Type:     sessionprotocol.MessageTypeStreamOpenOK,
		StreamID: connectRequest.StreamID,
	}); err != nil {
		return activePTYSession, nil, fmt.Errorf("failed to write sandbox tunnel stream.open acknowledgement: %w", err)
	}

	return activePTYSession, startPTYRelay(
		ctx,
		tunnelConn,
		activePTYSession,
		connectRequest.StreamID,
		relayResultCh,
	), nil
}

func startPTYRelay(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	session *ptySession,
	streamID int,
	relayResultCh chan<- activeTunnelStreamRelayResult,
) *activeTunnelStreamRelay {
	relay := &activeTunnelStreamRelay{
		PrimaryStreamID: streamID,
		ChannelKind:     sessionprotocol.ChannelKindPTY,
		MessageCh:       make(chan tunnelMessage),
	}

	go func() {
		result := activeTunnelStreamRelayResult{
			Relay:             relay,
			PTYSession:        session,
			UpdatesPTYSession: true,
		}
		if err := relayPTYSession(ctx, tunnelConn, session, streamID, relay.MessageCh); err != nil {
			result.Err = fmt.Errorf("sandbox tunnel pty relay failed: %w", err)
		}
		if session.IsExited() {
			result.PTYSession = nil
		}
		relayResultCh <- result
	}()

	return relay
}

func relayPTYSession(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	session *ptySession,
	streamID int,
	incomingMessages <-chan tunnelMessage,
) error {
	relayContext, cancel := context.WithCancel(ctx)
	defer cancel()

	attachedStreamIDs := map[int]struct{}{
		streamID: {},
	}
	sendWindowsByStreamID := map[int]*streamSendWindow{
		streamID: newStreamSendWindow(),
	}
	ptyOutputCh := make(chan []byte, 8)
	ptyOutputErrCh := make(chan error, 1)
	go readPTYOutput(session, ptyOutputCh, ptyOutputErrCh)

	for {
		select {
		case message := <-incomingMessages:
			switch message.MessageType {
			case websocket.MessageBinary:
				dataFrame, err := sessionprotocol.DecodeDataFrame(message.Payload)
				if err != nil {
					if writeErr := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
						Type:     sessionprotocol.MessageTypeStreamReset,
						StreamID: streamID,
						Code:     streamResetCodeInvalidStreamData,
						Message:  err.Error(),
					}); writeErr != nil {
						return fmt.Errorf("failed to write stream.reset for invalid pty data frame: %w", writeErr)
					}
					return nil
				}
				if _, isAttachedStream := attachedStreamIDs[int(dataFrame.StreamID)]; !isAttachedStream {
					if err := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
						Type:     sessionprotocol.MessageTypeStreamReset,
						StreamID: int(dataFrame.StreamID),
						Code:     streamResetCodeInvalidStreamData,
						Message:  fmt.Sprintf("stream data frame streamId %d is not attached to the active PTY session", dataFrame.StreamID),
					}); err != nil {
						return fmt.Errorf("failed to write stream.reset for mismatched pty data frame: %w", err)
					}
					return nil
				}
				if dataFrame.PayloadKind != sessionprotocol.PayloadKindRawBytes {
					if err := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
						Type:     sessionprotocol.MessageTypeStreamReset,
						StreamID: streamID,
						Code:     streamResetCodeInvalidStreamData,
						Message:  fmt.Sprintf("pty stream payloadKind %d is not supported", dataFrame.PayloadKind),
					}); err != nil {
						return fmt.Errorf("failed to write stream.reset for unsupported pty payload kind: %w", err)
					}
					return nil
				}
				if session.IsExited() {
					continue
				}
				if _, err := session.terminal.Write(dataFrame.Payload); err != nil {
					return fmt.Errorf("failed to write pty stdin payload: %w", err)
				}
				if err := writeStreamWindow(relayContext, tunnelConn, sessionprotocol.StreamWindow{
					Type:     sessionprotocol.MessageTypeStreamWindow,
					StreamID: int(dataFrame.StreamID),
					Bytes:    len(dataFrame.Payload),
				}); err != nil {
					return fmt.Errorf("failed to write stream.window for consumed pty data: %w", err)
				}
			case websocket.MessageText:
				controlMessageType, err := parseControlMessageType(message.Payload)
				if err == nil && controlMessageType == sessionprotocol.MessageTypeStreamWindow {
					streamWindow, windowErr := parseStreamWindow(message.Payload)
					if windowErr != nil {
						return windowErr
					}
					sendWindow := sendWindowsByStreamID[streamWindow.StreamID]
					if sendWindow == nil {
						if err := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
							Type:     sessionprotocol.MessageTypeStreamReset,
							StreamID: streamWindow.StreamID,
							Code:     streamResetCodeInvalidStreamData,
							Message:  fmt.Sprintf("stream.window streamId %d is not attached to the active PTY session", streamWindow.StreamID),
						}); err != nil {
							return fmt.Errorf("failed to write stream.reset for mismatched pty stream.window: %w", err)
						}
						return nil
					}
					if err := sendWindow.add(streamWindow.Bytes); err != nil {
						if writeErr := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
							Type:     sessionprotocol.MessageTypeStreamReset,
							StreamID: streamWindow.StreamID,
							Code:     streamResetCodeInvalidStreamWindow,
							Message:  err.Error(),
						}); writeErr != nil {
							return fmt.Errorf("failed to write stream.reset for excessive pty stream.window: %w", writeErr)
						}
						if streamWindow.StreamID != streamID {
							delete(attachedStreamIDs, streamWindow.StreamID)
							delete(sendWindowsByStreamID, streamWindow.StreamID)
							continue
						}
						return nil
					}
					continue
				}
				controlAction, err := handlePTYControlMessage(
					relayContext,
					tunnelConn,
					session,
					streamID,
					attachedStreamIDs,
					sendWindowsByStreamID,
					message.Payload,
				)
				if err != nil {
					return err
				}
				if controlAction == ptyControlActionDetachStream {
					continue
				}
				if controlAction == ptyControlActionCloseSession {
					return nil
				}
			default:
				return fmt.Errorf(
					"unsupported websocket message type for pty session: %s",
					message.MessageType.String(),
				)
			}
		case outputPayload := <-ptyOutputCh:
			for attachedStreamID := range attachedStreamIDs {
				sendWindow := sendWindowsByStreamID[attachedStreamID]
				if sendWindow == nil {
					continue
				}
				if !sendWindow.tryConsume(len(outputPayload)) {
					if err := writeStreamReset(relayContext, tunnelConn, sessionprotocol.StreamReset{
						Type:     sessionprotocol.MessageTypeStreamReset,
						StreamID: attachedStreamID,
						Code:     streamResetCodeStreamWindowExhausted,
						Message:  "pty stream send window is exhausted",
					}); err != nil {
						return fmt.Errorf("failed to write stream.reset for exhausted pty send window: %w", err)
					}
					delete(attachedStreamIDs, attachedStreamID)
					delete(sendWindowsByStreamID, attachedStreamID)
					if attachedStreamID == streamID {
						_ = session.CloseTerminal()
						return nil
					}
					continue
				}
				if err := writeBinaryDataFrame(
					relayContext,
					tunnelConn,
					uint32(attachedStreamID),
					sessionprotocol.PayloadKindRawBytes,
					outputPayload,
				); err != nil {
					return fmt.Errorf("failed to write pty output data frame: %w", err)
				}
			}
		case ptyOutputErr := <-ptyOutputErrCh:
			if ptyOutputErr == nil ||
				errors.Is(ptyOutputErr, io.EOF) ||
				errors.Is(ptyOutputErr, os.ErrClosed) ||
				errors.Is(ptyOutputErr, syscall.EIO) {
				continue
			}
			return fmt.Errorf("failed to read pty output: %w", ptyOutputErr)
		case <-session.exitedCh:
			for attachedStreamID := range attachedStreamIDs {
				if err := writeStreamEvent(relayContext, tunnelConn, sessionprotocol.StreamEvent{
					Type:     sessionprotocol.MessageTypeStreamEvent,
					StreamID: attachedStreamID,
					Event: sessionprotocol.PTYExitEvent{
						Type:     sessionprotocol.MessageTypePTYExit,
						ExitCode: session.ExitCode(),
					},
				}); err != nil {
					return fmt.Errorf("failed to write pty exit message: %w", err)
				}
			}
			_ = session.CloseTerminal()
			return nil
		}
	}
}

func readPTYOutput(session *ptySession, outputCh chan<- []byte, errCh chan<- error) {
	buffer := make([]byte, ptyOutputReadBufferBytes)
	for {
		readBytes, readErr := session.terminal.Read(buffer)
		if readBytes > 0 {
			chunk := make([]byte, readBytes)
			copy(chunk, buffer[:readBytes])
			outputCh <- chunk
		}
		if readErr != nil {
			errCh <- readErr
			return
		}
	}
}

func handlePTYControlMessage(
	ctx context.Context,
	tunnelConn *websocket.Conn,
	session *ptySession,
	streamID int,
	attachedStreamIDs map[int]struct{},
	sendWindowsByStreamID map[int]*streamSendWindow,
	payload []byte,
) (ptyControlAction, error) {
	messageType, err := parseControlMessageType(payload)
	if err != nil {
		return ptyControlActionContinue, fmt.Errorf("failed to parse pty control message type: %w", err)
	}

	switch messageType {
	case sessionprotocol.MessageTypeStreamOpen:
		connectRequest, err := parsePTYConnectRequest(payload)
		if err != nil {
			if writeErr := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
				Type:     sessionprotocol.MessageTypeStreamOpenError,
				StreamID: 0,
				Code:     connectErrorCodeInvalidConnectRequest,
				Message:  err.Error(),
			}); writeErr != nil {
				return ptyControlActionContinue, fmt.Errorf("failed to write stream.open.error during pty relay: %w", writeErr)
			}
			return ptyControlActionContinue, nil
		}

		if connectRequest.Channel.Session == sessionprotocol.PTYSessionModeCreate {
			if err := writeStreamOpenError(ctx, tunnelConn, sessionprotocol.StreamOpenError{
				Type:     sessionprotocol.MessageTypeStreamOpenError,
				StreamID: connectRequest.StreamID,
				Code:     connectErrorCodePTYSessionExists,
				Message:  "pty session already exists",
			}); err != nil {
				return ptyControlActionContinue, fmt.Errorf("failed to write stream.open.error during pty relay: %w", err)
			}
			return ptyControlActionContinue, nil
		}

		attachedStreamIDs[connectRequest.StreamID] = struct{}{}
		sendWindowsByStreamID[connectRequest.StreamID] = newStreamSendWindow()
		if err := writeStreamOpenOK(ctx, tunnelConn, sessionprotocol.StreamOpenOK{
			Type:     sessionprotocol.MessageTypeStreamOpenOK,
			StreamID: connectRequest.StreamID,
		}); err != nil {
			return ptyControlActionContinue, fmt.Errorf("failed to write stream.open.ok during pty relay: %w", err)
		}

		return ptyControlActionContinue, nil
	case sessionprotocol.MessageTypeStreamSignal:
		signalMessage, err := parsePTYResizeSignal(payload)
		if err != nil {
			if writeErr := writeStreamReset(ctx, tunnelConn, sessionprotocol.StreamReset{
				Type:     sessionprotocol.MessageTypeStreamReset,
				StreamID: streamID,
				Code:     streamResetCodeInvalidStreamSignal,
				Message:  err.Error(),
			}); writeErr != nil {
				return ptyControlActionCloseSession, fmt.Errorf("failed to write stream.reset for invalid pty signal: %w", writeErr)
			}
			return ptyControlActionCloseSession, nil
		}
		if _, isAttachedStream := attachedStreamIDs[signalMessage.StreamID]; !isAttachedStream {
			if err := writeStreamReset(ctx, tunnelConn, sessionprotocol.StreamReset{
				Type:     sessionprotocol.MessageTypeStreamReset,
				StreamID: signalMessage.StreamID,
				Code:     streamResetCodeInvalidStreamSignal,
				Message:  fmt.Sprintf("stream signal streamId %d is not attached to the active PTY session", signalMessage.StreamID),
			}); err != nil {
				return ptyControlActionCloseSession, fmt.Errorf("failed to write stream.reset for mismatched pty signal: %w", err)
			}
			return ptyControlActionCloseSession, nil
		}
		if err := session.Resize(signalMessage.Signal.Cols, signalMessage.Signal.Rows); err != nil {
			return ptyControlActionContinue, fmt.Errorf("failed to resize pty session: %w", err)
		}
		return ptyControlActionContinue, nil
	case sessionprotocol.MessageTypeStreamClose:
		closeRequest, err := parseStreamClose(payload)
		if err != nil {
			if writeErr := writeStreamReset(ctx, tunnelConn, sessionprotocol.StreamReset{
				Type:     sessionprotocol.MessageTypeStreamReset,
				StreamID: streamID,
				Code:     streamResetCodeInvalidStreamClose,
				Message:  err.Error(),
			}); writeErr != nil {
				return ptyControlActionCloseSession, fmt.Errorf("failed to write stream.reset for invalid pty close: %w", writeErr)
			}
			return ptyControlActionCloseSession, nil
		}
		if _, isAttachedStream := attachedStreamIDs[closeRequest.StreamID]; !isAttachedStream {
			if err := writeStreamReset(ctx, tunnelConn, sessionprotocol.StreamReset{
				Type:     sessionprotocol.MessageTypeStreamReset,
				StreamID: closeRequest.StreamID,
				Code:     streamResetCodeInvalidStreamClose,
				Message:  fmt.Sprintf("stream close streamId %d is not attached to the active PTY session", closeRequest.StreamID),
			}); err != nil {
				return ptyControlActionCloseSession, fmt.Errorf("failed to write stream.reset for mismatched pty close: %w", err)
			}
			return ptyControlActionCloseSession, nil
		}
		if closeRequest.StreamID != streamID {
			delete(attachedStreamIDs, closeRequest.StreamID)
			delete(sendWindowsByStreamID, closeRequest.StreamID)
			return ptyControlActionDetachStream, nil
		}

		exitCode, terminateErr := session.Terminate()
		if terminateErr != nil {
			if writeErr := writeStreamReset(ctx, tunnelConn, sessionprotocol.StreamReset{
				Type:     sessionprotocol.MessageTypeStreamReset,
				StreamID: streamID,
				Code:     streamResetCodeStreamCloseFailed,
				Message:  terminateErr.Error(),
			}); writeErr != nil {
				return ptyControlActionCloseSession, fmt.Errorf("failed to write stream.reset for pty close failure: %w", writeErr)
			}
			return ptyControlActionCloseSession, nil
		}

		for attachedStreamID := range attachedStreamIDs {
			if err := writeStreamEvent(ctx, tunnelConn, sessionprotocol.StreamEvent{
				Type:     sessionprotocol.MessageTypeStreamEvent,
				StreamID: attachedStreamID,
				Event: sessionprotocol.PTYExitEvent{
					Type:     sessionprotocol.MessageTypePTYExit,
					ExitCode: exitCode,
				},
			}); err != nil {
				return ptyControlActionCloseSession, fmt.Errorf("failed to write stream.event for pty close: %w", err)
			}
		}

		_ = session.CloseTerminal()
		return ptyControlActionCloseSession, nil
	default:
		return ptyControlActionContinue, fmt.Errorf("unsupported pty control message type '%s'", messageType)
	}
}

func parsePTYResizeSignal(payload []byte) (sessionprotocol.StreamSignal, error) {
	var signalMessage sessionprotocol.StreamSignal
	if err := json.Unmarshal(payload, &signalMessage); err != nil {
		return sessionprotocol.StreamSignal{}, fmt.Errorf("stream.signal must be valid JSON: %w", err)
	}

	if signalMessage.Type != sessionprotocol.MessageTypeStreamSignal {
		return sessionprotocol.StreamSignal{}, fmt.Errorf("stream.signal request type must be '%s'", sessionprotocol.MessageTypeStreamSignal)
	}
	if signalMessage.StreamID <= 0 {
		return sessionprotocol.StreamSignal{}, fmt.Errorf("stream.signal request streamId must be a positive integer")
	}
	if signalMessage.Signal.Type != sessionprotocol.MessageTypePTYResize {
		return sessionprotocol.StreamSignal{}, fmt.Errorf("stream.signal signal.type must be '%s'", sessionprotocol.MessageTypePTYResize)
	}
	if signalMessage.Signal.Cols < 1 || signalMessage.Signal.Rows < 1 {
		return sessionprotocol.StreamSignal{}, fmt.Errorf("pty resize signal cols and rows must be greater than or equal to 1")
	}
	if signalMessage.Signal.Cols > 65535 || signalMessage.Signal.Rows > 65535 {
		return sessionprotocol.StreamSignal{}, fmt.Errorf("pty resize signal cols and rows must be less than or equal to 65535")
	}

	return signalMessage, nil
}

func parseStreamClose(payload []byte) (sessionprotocol.StreamClose, error) {
	var closeRequest sessionprotocol.StreamClose
	if err := json.Unmarshal(payload, &closeRequest); err != nil {
		return sessionprotocol.StreamClose{}, fmt.Errorf("stream.close must be valid JSON: %w", err)
	}

	if closeRequest.Type != sessionprotocol.MessageTypeStreamClose {
		return sessionprotocol.StreamClose{}, fmt.Errorf("stream.close request type must be '%s'", sessionprotocol.MessageTypeStreamClose)
	}
	if closeRequest.StreamID <= 0 {
		return sessionprotocol.StreamClose{}, fmt.Errorf("stream.close request streamId must be a positive integer")
	}

	return closeRequest, nil
}

File: apps/sandbox-runtime/src/tunnel/pty-channel.ts
import {
  decodeDataFrame,
  PayloadKindRawBytes,
  type StreamOpen,
} from "@mistle/sandbox-session-protocol";
import type WebSocket from "ws";

import { createAbortRace, ignorePromiseRejectionAfterAbort } from "./abortable-race.js";
import type { ActiveTunnelStreamRelay, ActiveTunnelStreamRelayResult } from "./active-relay.js";
import { AsyncQueue } from "./async-queue.js";
import type { TunnelSocketMessage } from "./connect-request.js";
import {
  parseControlMessageType,
  parsePtyConnectRequest,
  parsePtyResizeSignal,
  parseStreamCloseMessage,
} from "./connect-request.js";
import {
  CONNECT_ERROR_CODE_INVALID_CONNECT_REQUEST,
  CONNECT_ERROR_CODE_PTY_SESSION_CREATE_FAILED,
  CONNECT_ERROR_CODE_PTY_SESSION_EXISTS,
  CONNECT_ERROR_CODE_PTY_SESSION_UNAVAILABLE,
  STREAM_RESET_CODE_INVALID_STREAM_CLOSE,
  STREAM_RESET_CODE_INVALID_STREAM_DATA,
  STREAM_RESET_CODE_INVALID_STREAM_SIGNAL,
  STREAM_RESET_CODE_INVALID_STREAM_WINDOW,
  STREAM_RESET_CODE_STREAM_CLOSE_FAILED,
  STREAM_RESET_CODE_STREAM_WINDOW_EXHAUSTED,
  createPtyExitEventMessage,
  writeBinaryDataFrame,
  writeStreamEvent,
  writeStreamOpenError,
  writeStreamOpenOk,
  writeStreamReset,
  writeStreamWindow,
} from "./messages.js";
import { PtySession, PtySessionOutputClosedError, startPtySession } from "./pty-session.js";
import { StreamSendWindow } from "./stream-window.js";

async function handlePtyControlMessage(input: {
  signal: AbortSignal;
  tunnelSocket: WebSocket;
  session: PtySession;
  primaryStreamId: number;
  attachedStreamIds: Set<number>;
  sendWindowsByStreamId: Map<number, StreamSendWindow>;
  payload: string;
}): Promise<"continue" | "detach-stream" | "close-session"> {
  const messageType = parseControlMessageType(input.payload);

  switch (messageType) {
    case "stream.open": {
      const connectRequest = parsePtyConnectRequest(input.payload);
      if (connectRequest.channel.kind !== "pty") {
        throw new Error("pty stream.open request channel.kind must be 'pty'");
      }
      if (connectRequest.channel.session === "create") {
        await writeStreamOpenError(input.tunnelSocket, {
          type: "stream.open.error",
          streamId: connectRequest.streamId,
          code: CONNECT_ERROR_CODE_PTY_SESSION_EXISTS,
          message: "pty session already exists",
        });
        return "continue";
      }

      input.attachedStreamIds.add(connectRequest.streamId);
      input.sendWindowsByStreamId.set(connectRequest.streamId, new StreamSendWindow());
      await writeStreamOpenOk(input.tunnelSocket, {
        type: "stream.open.ok",
        streamId: connectRequest.streamId,
      });
      return "continue";
    }
    case "stream.signal": {
      const signalMessage = parsePtyResizeSignal(input.payload);
      if (!input.attachedStreamIds.has(signalMessage.streamId)) {
        await writeStreamReset(input.tunnelSocket, {
          type: "stream.reset",
          streamId: signalMessage.streamId,
          code: STREAM_RESET_CODE_INVALID_STREAM_SIGNAL,
          message: `stream signal streamId ${String(signalMessage.streamId)} is not attached to the active PTY session`,
        });
        return "close-session";
      }

      input.session.resize(signalMessage.signal.cols, signalMessage.signal.rows);
      return "continue";
    }
    case "stream.close": {
      const closeMessage = parseStreamCloseMessage(input.payload);
      if (!input.attachedStreamIds.has(closeMessage.streamId)) {
        await writeStreamReset(input.tunnelSocket, {
          type: "stream.reset",
          streamId: closeMessage.streamId,
          code: STREAM_RESET_CODE_INVALID_STREAM_CLOSE,
          message: `stream close streamId ${String(closeMessage.streamId)} is not attached to the active PTY session`,
        });
        return "close-session";
      }
      if (closeMessage.streamId !== input.primaryStreamId) {
        input.attachedStreamIds.delete(closeMessage.streamId);
        input.sendWindowsByStreamId.delete(closeMessage.streamId);
        return "detach-stream";
      }

      let exitCode: number;
      try {
        exitCode = await input.session.terminate();
      } catch (error) {
        await writeStreamReset(input.tunnelSocket, {
          type: "stream.reset",
          streamId: input.primaryStreamId,
          code: STREAM_RESET_CODE_STREAM_CLOSE_FAILED,
          message: error instanceof Error ? error.message : String(error),
        });
        return "close-session";
      }

      for (const attachedStreamId of input.attachedStreamIds) {
        await writeStreamEvent(
          input.tunnelSocket,
          createPtyExitEventMessage(attachedStreamId, exitCode),
        );
      }
      return "close-session";
    }
    default:
      throw new Error(`unsupported pty control message type '${messageType}'`);
  }
}

async function relayPtySession(input: {
  signal: AbortSignal;
  tunnelSocket: WebSocket;
  session: PtySession;
  primaryStreamId: number;
  messages: AsyncQueue<TunnelSocketMessage>;
}): Promise<PtySession | undefined> {
  const attachedStreamIds = new Set<number>([input.primaryStreamId]);
  const sendWindowsByStreamId = new Map<number, StreamSendWindow>([
    [input.primaryStreamId, new StreamSendWindow()],
  ]);
  const exitPromise = input.session.waitForExit();
  let pendingExitCode: number | undefined;

  while (!input.signal.aborted) {
    const nextMessageAbortController = new AbortController();
    const nextOutputAbortController = new AbortController();
    const abortRace = createAbortRace(input.signal);
    const nextExit =
      pendingExitCode === undefined
        ? exitPromise.then((exitCode) => ({
            source: "exit" as const,
            exitCode,
          }))
        : new Promise<never>(() => undefined);

    let event:
      | {
          source: "message";
          message: TunnelSocketMessage;
        }
      | {
          source: "output";
          output: Uint8Array;
        }
      | {
          source: "output-closed";
        }
      | {
          source: "exit";
          exitCode: number;
        };

    try {
      const nextMessage = ignorePromiseRejectionAfterAbort(
        input.messages.next(nextMessageAbortController.signal).then((message) => ({
          source: "message" as const,
          message,
        })),
        nextMessageAbortController.signal,
      );
      const nextOutput = ignorePromiseRejectionAfterAbort(
        input.session
          .nextOutput(nextOutputAbortController.signal)
          .then((output) => ({
            source: "output" as const,
            output,
          }))
          .catch((error: unknown) => {
            if (error instanceof PtySessionOutputClosedError) {
              return {
                source: "output-closed" as const,
              };
            }

            throw error;
          }),
        nextOutputAbortController.signal,
      );

      event = await Promise.race([nextMessage, nextOutput, nextExit, abortRace.promise]);
    } finally {
      abortRace.dispose();
      nextMessageAbortController.abort();
      nextOutputAbortController.abort();
    }

    if (event.source === "output") {
      for (const attachedStreamId of attachedStreamIds) {
        const sendWindow = sendWindowsByStreamId.get(attachedStreamId);
        if (sendWindow === undefined) {
          continue;
        }
        if (!sendWindow.tryConsume(event.output.length)) {
          await writeStreamReset(input.tunnelSocket, {
            type: "stream.reset",
            streamId: attachedStreamId,
            code: STREAM_RESET_CODE_STREAM_WINDOW_EXHAUSTED,
            message: "pty stream send window is exhausted",
          });
          attachedStreamIds.delete(attachedStreamId);
          sendWindowsByStreamId.delete(attachedStreamId);
          if (attachedStreamId === input.primaryStreamId) {
            await input.session.terminate().catch(() => undefined);
            return undefined;
          }
          continue;
        }

        await writeBinaryDataFrame(input.tunnelSocket, {
          streamId: attachedStreamId,
          payloadKind: PayloadKindRawBytes,
          payload: event.output,
        });
      }
      continue;
    }

    if (event.source === "output-closed") {
      const exitCode = pendingExitCode ?? (await exitPromise);
      for (const attachedStreamId of attachedStreamIds) {
        await writeStreamEvent(
          input.tunnelSocket,
          createPtyExitEventMessage(attachedStreamId, exitCode),
        );
      }
      return undefined;
    }

    if (event.source === "exit") {
      pendingExitCode = event.exitCode;
      continue;
    }

    const message = event.message;
    if (message.kind === "binary") {
      let dataFrame;
      try {
        dataFrame = decodeDataFrame(message.payload);
      } catch (error) {
        await writeStreamReset(input.tunnelSocket, {
          type: "stream.reset",
          streamId: input.primaryStreamId,
          code: STREAM_RESET_CODE_INVALID_STREAM_DATA,
          message: error instanceof Error ? error.message : String(error),
        });
        return input.session;
      }

      if (!attachedStreamIds.has(dataFrame.streamId)) {
        await writeStreamReset(input.tunnelSocket, {
          type: "stream.reset",
          streamId: dataFrame.streamId,
          code: STREAM_RESET_CODE_INVALID_STREAM_DATA,
          message: `stream data frame streamId ${String(dataFrame.streamId)} is not attached to the active PTY session`,
        });
        return input.session;
      }
      if (dataFrame.payloadKind !== PayloadKindRawBytes) {
        await writeStreamReset(input.tunnelSocket, {
          type: "stream.reset",
          streamId: input.primaryStreamId,
          code: STREAM_RESET_CODE_INVALID_STREAM_DATA,
          message: `pty stream payloadKind ${String(dataFrame.payloadKind)} is not supported`,
        });
        return input.session;
      }

      input.session.write(dataFrame.payload);
      await writeStreamWindow(input.tunnelSocket, {
        type: "stream.window",
        streamId: dataFrame.streamId,
        bytes: dataFrame.payload.length,
      });
      continue;
    }

    const controlMessageType = parseControlMessageType(message.payload);
    if (controlMessageType === "stream.window") {
      const parsedWindow = JSON.parse(message.payload);
      const streamId =
        typeof parsedWindow.streamId === "number" && Number.isInteger(parsedWindow.streamId)
          ? parsedWindow.streamId
          : 0;
      const bytes =
        typeof parsedWindow.bytes === "number" && Number.isInteger(parsedWindow.bytes)
          ? parsedWindow.bytes
          : 0;
      const sendWindow = sendWindowsByStreamId.get(streamId);
      if (sendWindow === undefined) {
        await writeStreamReset(input.tunnelSocket, {
          type: "stream.reset",
          streamId,
          code: STREAM_RESET_CODE_INVALID_STREAM_DATA,
          message: `stream.window streamId ${String(streamId)} is not attached to the active PTY session`,
        });
        return input.session;
      }
      try {
        sendWindow.add(bytes);
      } catch (error) {
        await writeStreamReset(input.tunnelSocket, {
          type: "stream.reset",
          streamId,
          code: STREAM_RESET_CODE_INVALID_STREAM_WINDOW,
          message: error instanceof Error ? error.message : String(error),
        });
        if (streamId !== input.primaryStreamId) {
          attachedStreamIds.delete(streamId);
          sendWindowsByStreamId.delete(streamId);
          continue;
        }
        return input.session;
      }
      continue;
    }

    const controlAction = await handlePtyControlMessage({
      signal: input.signal,
      tunnelSocket: input.tunnelSocket,
      session: input.session,
      primaryStreamId: input.primaryStreamId,
      attachedStreamIds,
      sendWindowsByStreamId,
      payload: message.payload,
    });
    if (controlAction === "close-session") {
      return input.session.isExited() ? undefined : input.session;
    }
  }

  return input.session;
}

export async function handlePtyConnectRequest(input: {
  signal: AbortSignal;
  tunnelSocket: WebSocket;
  rawPayload: string;
  streamId: number;
  activePtySession: PtySession | undefined;
  relayResultQueue: AsyncQueue<ActiveTunnelStreamRelayResult>;
}): Promise<{
  ptySession?: PtySession;
  relay?: ActiveTunnelStreamRelay;
}> {
  let connectRequest: StreamOpen;
  try {
    connectRequest = parsePtyConnectRequest(input.rawPayload);
  } catch (error) {
    await writeStreamOpenError(input.tunnelSocket, {
      type: "stream.open.error",
      streamId: input.streamId,
      code: CONNECT_ERROR_CODE_INVALID_CONNECT_REQUEST,
      message: error instanceof Error ? error.message : String(error),
    });
    if (input.activePtySession === undefined) {
      return {};
    }

    return { ptySession: input.activePtySession };
  }

  if (connectRequest.channel.kind !== "pty") {
    throw new Error("pty stream.open request channel.kind must be 'pty'");
  }

  let activePtySession = input.activePtySession;
  if (connectRequest.channel.session === "create") {
    if (activePtySession !== undefined && !activePtySession.isExited()) {
      await writeStreamOpenError(input.tunnelSocket, {
        type: "stream.open.error",
        streamId: input.streamId,
        code: CONNECT_ERROR_CODE_PTY_SESSION_EXISTS,
        message: "pty session already exists",
      });
      if (activePtySession === undefined) {
        return {};
      }

      return { ptySession: activePtySession };
    }

    try {
      activePtySession = startPtySession(connectRequest);
    } catch (error) {
      await writeStreamOpenError(input.tunnelSocket, {
        type: "stream.open.error",
        streamId: input.streamId,
        code: CONNECT_ERROR_CODE_PTY_SESSION_CREATE_FAILED,
        message: error instanceof Error ? error.message : String(error),
      });
      if (input.activePtySession === undefined) {
        return {};
      }

      return { ptySession: input.activePtySession };
    }
  }

  if (activePtySession === undefined || activePtySession.isExited()) {
    await writeStreamOpenError(input.tunnelSocket, {
      type: "stream.open.error",
      streamId: input.streamId,
      code: CONNECT_ERROR_CODE_PTY_SESSION_UNAVAILABLE,
      message: "pty session is not available",
    });
    if (activePtySession === undefined) {
      return {};
    }

    return { ptySession: activePtySession };
  }

  await writeStreamOpenOk(input.tunnelSocket, {
    type: "stream.open.ok",
    streamId: input.streamId,
  });

  const relay: ActiveTunnelStreamRelay = {
    primaryStreamId: input.streamId,
    channelKind: "pty",
    messages: new AsyncQueue<TunnelSocketMessage>(),
  };

  void relayPtySession({
    signal: input.signal,
    tunnelSocket: input.tunnelSocket,
    session: activePtySession,
    primaryStreamId: input.streamId,
    messages: relay.messages,
  })
    .then((ptySession) => {
      input.relayResultQueue.push({
        relay,
        ptySession,
        updatesPtySession: true,
      });
    })
    .catch((error: unknown) => {
      input.relayResultQueue.push({
        relay,
        error: error instanceof Error ? error : new Error(String(error)),
        ptySession: activePtySession,
        updatesPtySession: true,
      });
    });

  if (activePtySession === undefined) {
    return { relay };
  }

  return {
    ptySession: activePtySession,
    relay,
  };
}
